jupyter_sphinx/execute.py

```diff
@@ -0,0 +1,373 @@
+    ))
+
+
+def sphinx_abs_dir(env):
+    # We write the output files into
+    # output_directory / jupyter_execute / path relative to source directory
+    # Sphinx expects download links relative to source file or relative to
+    # source dir and prepended with '/'. We use the latter option.
+    return '/' + os.path.relpath(
+        os.path.abspath(os.path.join(
+            output_directory(env),
+            os.path.dirname(env.docname),
+        )),
+        os.path.abspath(env.app.srcdir)
+    )
+
+
+class ExecuteJupyterCells(SphinxTransform):
+    default_priority = 180  # An early transform, idk
+
+    def apply(self):
+        doctree = self.document
+        doc_relpath = os.path.dirname(self.env.docname)  # relative to src dir
+        docname = os.path.basename(self.env.docname)
+        default_kernel = self.config.jupyter_execute_default_kernel
+        default_names = default_notebook_names(docname)
+
+        # Check if we have anything to execute.
+        if not doctree.traverse(Cell):
+            return
+
+        logger.info('executing {}'.format(docname))
+        output_dir = os.path.join(output_directory(self.env), doc_relpath)
+
+        # Start new notebook whenever a cell has 'new_notebook' specified
+        nodes_by_notebook = split_on(
+            itemgetter('new_notebook'),
+            doctree.traverse(Cell)
+        )
+
+        for nodes in nodes_by_notebook:
+            kernel_name = nodes[0]['kernel_name'] or default_kernel
+            notebook_name = nodes[0]['notebook_name'] or next(default_names)
+
+            notebook = execute_cells(
+                kernel_name,
+                [nbformat.v4.new_code_cell(node.astext()) for node in nodes],
+                self.config.jupyter_execute_kwargs,
+            )
+            # Modifies 'notebook' in-place, adding metadata specifying the
+            # filenames of the saved outputs.
+            write_notebook_output(notebook, output_dir, notebook_name)
+            # Add doctree nodes for cell output; images reference the filenames
+            # we just wrote to; sphinx copies these when writing outputs.
+            for node, cell in zip(nodes, notebook.cells):
+                output_nodes = cell_output_to_nodes(
+                    cell,
+                    self.config.jupyter_execute_data_priority,
+                    sphinx_abs_dir(self.env)
+                )
+                attach_outputs(output_nodes, node)
+
+
+def setup(app):
+    # Configuration
+    app.add_config_value(
+        'jupyter_execute_kwargs',
+        dict(timeout=-1, allow_errors=True),
+        'env'
+    )
+    app.add_config_value(
+        'jupyter_execute_default_kernel',
+        'python3',
+        'env'
+    )
+    app.add_config_value(
+        'jupyter_execute_data_priority',
+        [
+            'text/html',
+            'image/svg+xml',
+            'image/png',
+            'image/jpeg',
+            'text/latex',
+            'text/plain'
```

Comment:
What takes care of the rendering the script? Extra page header? What if we want to render something also in latex output?

Is it a standard behavior for various types of data? I am particularly interested in plotly, but they seem to not have a renderer that would function this way.

Reply:
There is a library called the HTML embedder which can be included as a script tag. This is what we use to render widgets in e.g. http://nbviewer.jupyter.org/github/jupyter-widgets/ipywidgets/blob/master/docs/source/examples/Widget%20List.ipynb

Although, it will also require the widgets state to be stored in a separate `application/vnd.jupyter.widget-view+json` script tag. I will probably open a PR adding this later.

Note that plotly is now based on Jupyter widgets (since 3.0) so if they did things right, this will also enable the rendering of plotly charts.

Reply:
They have `FigureWidget`, but it doesn't seem to save the actual figure data in the widget state right now. (This should go into `widget_state`, right?) They also support bundling json with the output.

Reply:
Do you have a link to the HTML embedder library? I failed to find it after a quick google search.

Reply:
The name of the package is `@jupyter-widgets/html-manager` 
 and the code is in the ipywidgets repository. You can read more about it here:

https://ipywidgets.readthedocs.io/en/stable/embedding.html

Reply:
Would this also work with scripts that have `src` tag set? Some widgets will contain a lot of data, and extracting them into a separate file would improve the page load times.

Thinking of producing output other than html: I remember there was a way to render a widget into a static image, is that right? Does it work on custom widgets? Can this be done without a browser (e.g. via an electron app)?

Reply:
There is now https://github.com/jbweston/jupyter-sphinx/pull/2 where I've started implementing ipywidgets support. Perhaps we should first merge this PR and then we can address that one (there are still several unanswered questions.